File: backend/app/api/routes/claims.py
import json
from typing import Any
from datetime import datetime

from fastapi import APIRouter, Depends, Query, Header, HTTPException
from sqlalchemy.orm import Session

from app.db.database import get_db
from app import models
from app.utils.sorting import parse_sort_params
from app.schemas.claims import ClaimOut, ClaimListQuery, ClaimListResponse
from app.services import risk_engine, pipeline

router = APIRouter(tags=["claims"])


def get_current_clinic(x_clinic_id: int | None = Header(None)) -> int:
    # Enforce clinic-level isolation via header `X-Clinic-Id` for now.
    if x_clinic_id is None:
        # default to clinic 1 for local dev
        return 1
    return x_clinic_id


@router.get("/claims", response_model=ClaimListResponse)
def list_claims(
    page: int = Query(1, ge=1),
    size: int = Query(25, ge=1, le=200),
    status: str | None = Query(None),
    risk_level: str | None = Query(None),
    payer_id: int | None = Query(None),
    start_date: datetime | None = Query(None),
    end_date: datetime | None = Query(None),
    search: str | None = Query(None),
    sort_by: str | None = Query(None),
    sort_dir: str | None = Query(None),
    clinic_id: int = Depends(get_current_clinic),
    db: Session = Depends(get_db),
) -> list[ClaimOut]:
    base_q = db.query(models.Claim).filter(models.Claim.clinic_id == clinic_id)

    if status:
        base_q = base_q.filter(models.Claim.status == status)
    if risk_level:
        base_q = base_q.filter(models.Claim.risk_level == risk_level)
    if payer_id:
        base_q = base_q.filter(models.Claim.payer_id == payer_id)
    if start_date:
        base_q = base_q.filter(models.Claim.created_at >= start_date)
    if end_date:
        base_q = base_q.filter(models.Claim.created_at <= end_date)
    if search:
        base_q = base_q.filter(models.Claim.claim_id.ilike(f"%{search}%"))

    total = base_q.count()

    sort_list = parse_sort_params(sort_by, sort_dir)

    sort_map = {
        'created_at': models.Claim.created_at,
        'amount': models.Claim.amount,
        'status': models.Claim.status,
        'risk_score': models.Claim.risk_score,
        'claim_id': models.Claim.claim_id,
    }

    applied_any = False
    for col_name, dir_ in sort_list:
        if col_name in sort_map:
            col = sort_map[col_name]
            if dir_ == 'asc':
                base_q = base_q.order_by(col.asc())
            else:
                base_q = base_q.order_by(col.desc())
            applied_any = True

    if not applied_any:
        base_q = base_q.order_by(models.Claim.created_at.desc())

    q = base_q.offset((page - 1) * size).limit(size)
    results = []
    for c in q.all():
        results.append({
            "id": c.id,
            "claim_id": c.claim_id,
            "clinic_id": c.clinic_id,
            "payer_id": c.payer_id,
            "payer_name": getattr(c.payer, "name", None),
            "patient_id": getattr(c, "patient_id", None),
            "patient_name": f"{getattr(c.patient, 'first_name', '')} {getattr(c.patient, 'last_name', '')}".strip() or None,
            "amount": float(c.amount),
            "status": c.status,
            "risk_level": c.risk_level,
            "risk_score": c.risk_score,
            "created_at": c.created_at,
            "updated_at": c.updated_at,
        })
    return {"items": results, "total": total, "page": page, "size": size}


@router.get("/claims/{claim_id}", response_model=ClaimOut)
def get_claim(claim_id: str, clinic_id: int = Depends(get_current_clinic), db: Session = Depends(get_db)) -> Any:
    claim = db.query(models.Claim).filter(models.Claim.claim_id == claim_id, models.Claim.clinic_id == clinic_id).first()
    if not claim:
        raise HTTPException(status_code=404, detail="Claim not found")
    return {
        "id": claim.id,
        "claim_id": claim.claim_id,
        "clinic_id": claim.clinic_id,
        "payer_id": claim.payer_id,
        "payer_name": getattr(claim.payer, "name", None),
        "patient_id": getattr(claim, "patient_id", None),
        "patient_name": f"{getattr(claim.patient, 'first_name', '')} {getattr(claim.patient, 'last_name', '')}".strip() or None,
        "amount": float(claim.amount),
        "status": claim.status,
        "risk_level": claim.risk_level,
        "risk_score": claim.risk_score,
        "created_at": claim.created_at,
        "updated_at": claim.updated_at,
    }




@router.get('/claims/{claim_id}/analyze')
def analyze_claim(claim_id: str, clinic_id: int = Depends(get_current_clinic), db: Session = Depends(get_db)):
    claim = db.query(models.Claim).filter(models.Claim.claim_id == claim_id, models.Claim.clinic_id == clinic_id).first()
    if not claim:
        raise HTTPException(status_code=404, detail='Claim not found')
    issues = risk_engine.analyze_and_persist(db, claim)
    return {'issues': [issue.__dict__ if hasattr(issue, '__dict__') else issue for issue in issues], 'risk_level': claim.risk_level, 'risk_score': claim.risk_score}


def _get_claim_or_404(claim_id: str, clinic_id: int, db: Session) -> models.Claim:
    claim = db.query(models.Claim).filter(models.Claim.claim_id == claim_id, models.Claim.clinic_id == clinic_id).first()
    if not claim:
        raise HTTPException(status_code=404, detail="Claim not found")
    return claim


def _recommendation_out(rec: models.Recommendation) -> dict:
    return {
        "id": rec.id,
        "claim_id": rec.claim_id,
        "action_type": rec.action_type,
        "rationale": rec.note,
        "confidence": rec.confidence,
        "low_confidence": bool(rec.low_confidence),
        "cited_issue_types": json.loads(rec.cited_issue_types) if rec.cited_issue_types else [],
        "secondary_options": json.loads(rec.secondary_options) if rec.secondary_options else [],
        "approval_status": rec.approval_status,
        "created_at": rec.created_at,
        "decided_at": rec.decided_at,
    }


def _activity_out(log: models.ActivityLog) -> dict:
    return {
        "id": log.id,
        "action": log.action,
        "details": json.loads(log.details) if log.details else None,
        "created_at": log.created_at,
    }


def _decision_out(decision) -> dict:
    return {"decision": decision.decision, "reason_code": decision.reason_code, "rule": decision.rule}


def _execution_out(execution) -> dict | None:
    if execution is None:
        return None
    kind = type(execution).__name__  # "FollowUpResult" | "FollowUpFailure" | "ReminderResult" | "ReminderFailure"
    out = {"kind": kind, **execution.__dict__}
    if "due_at" in out and out["due_at"] is not None:
        out["due_at"] = out["due_at"].isoformat()
    return out


@router.get("/claims/{claim_id}/recommendation")
def get_latest_recommendation(claim_id: str, clinic_id: int = Depends(get_current_clinic), db: Session = Depends(get_db)):
    claim = _get_claim_or_404(claim_id, clinic_id, db)
    rec = (
        db.query(models.Recommendation)
        .filter(models.Recommendation.claim_id == claim.id)
        .order_by(models.Recommendation.created_at.desc(), models.Recommendation.id.desc())
        .first()
    )
    return {"recommendation": _recommendation_out(rec) if rec else None}


@router.post("/claims/{claim_id}/recommendation")
def generate_recommendation(claim_id: str, clinic_id: int = Depends(get_current_clinic), db: Session = Depends(get_db)):
    claim = _get_claim_or_404(claim_id, clinic_id, db)
    outcome = pipeline.generate_recommendation(db, claim)
    return {
        "stage": outcome.stage,
        "detail": outcome.detail,
        "decision": _decision_out(outcome.decision) if outcome.decision else None,
        "recommendation": _recommendation_out(outcome.recommendation) if outcome.recommendation else None,
        "escalation_id": outcome.escalation.escalation_id if outcome.escalation else None,
    }


@router.post("/claims/{claim_id}/recommendation/{recommendation_id}/approve")
def approve_recommendation(
    claim_id: str,
    recommendation_id: int,
    clinic_id: int = Depends(get_current_clinic),
    db: Session = Depends(get_db),
):
    claim = _get_claim_or_404(claim_id, clinic_id, db)
    rec = (
        db.query(models.Recommendation)
        .filter(models.Recommendation.id == recommendation_id, models.Recommendation.claim_id == claim.id)
        .first()
    )
    if not rec:
        raise HTTPException(status_code=404, detail="Recommendation not found")
    if rec.approval_status != pipeline.STATUS_PENDING:
        raise HTTPException(status_code=409, detail=f"Recommendation is not pending approval (status={rec.approval_status})")

    outcome = pipeline.decide_recommendation(db, claim, rec, approved=True)
    return {
        "recommendation": _recommendation_out(outcome.recommendation),
        "decision": _decision_out(outcome.decision),
        "execution": _execution_out(outcome.execution),
        "execution_decision": _decision_out(outcome.execution_decision) if outcome.execution_decision else None,
        "escalation_id": outcome.escalation.escalation_id if outcome.escalation else None,
        "activity": _activity_out(outcome.activity),
    }


@router.post("/claims/{claim_id}/recommendation/{recommendation_id}/decline")
def decline_recommendation(
    claim_id: str,
    recommendation_id: int,
    clinic_id: int = Depends(get_current_clinic),
    db: Session = Depends(get_db),
):
    claim = _get_claim_or_404(claim_id, clinic_id, db)
    rec = (
        db.query(models.Recommendation)
        .filter(models.Recommendation.id == recommendation_id, models.Recommendation.claim_id == claim.id)
        .first()
    )
    if not rec:
        raise HTTPException(status_code=404, detail="Recommendation not found")
    if rec.approval_status != pipeline.STATUS_PENDING:
        raise HTTPException(status_code=409, detail=f"Recommendation is not pending approval (status={rec.approval_status})")

    outcome = pipeline.decide_recommendation(db, claim, rec, approved=False)
    return {
        "recommendation": _recommendation_out(outcome.recommendation),
        "decision": _decision_out(outcome.decision),
        "execution": _execution_out(outcome.execution),
        "execution_decision": _decision_out(outcome.execution_decision) if outcome.execution_decision else None,
        "escalation_id": outcome.escalation.escalation_id if outcome.escalation else None,
        "activity": _activity_out(outcome.activity),
    }


@router.get("/claims/{claim_id}/activity")
def get_claim_activity(claim_id: str, clinic_id: int = Depends(get_current_clinic), db: Session = Depends(get_db)):
    claim = _get_claim_or_404(claim_id, clinic_id, db)
    logs = (
        db.query(models.ActivityLog)
        .filter(models.ActivityLog.claim_id == claim.id)
        .order_by(models.ActivityLog.created_at.desc(), models.ActivityLog.id.desc())
        .all()
    )
    return {"items": [_activity_out(log) for log in logs]}

File: frontend/src/services/claims.ts
export type Claim = {
  id: number
  claim_id: string
  clinic_id: number
  payer_id: number
  payer_name?: string
  patient_id?: number
  patient_name?: string
  amount: number
  status: string
  risk_level: string
  risk_score: number
  created_at: string
  updated_at?: string
}

const BASE = (import.meta as any).env?.VITE_API_BASE ?? 'http://127.0.0.1:8000'

export type ClaimListResponse = {
  items: Claim[]
  total: number
  page: number
  size: number
}

export async function fetchClaims(params: Record<string, string | number | undefined> = {}) {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null) query.set(k, String(v))
  })
  const url = `${BASE}/api/claims?${query.toString()}`
  const res = await fetch(url)
  if (!res.ok) throw new Error('Failed to fetch claims')
  return (await res.json()) as ClaimListResponse
}

export async function fetchClaim(id: string) {
  const res = await fetch(`${BASE}/api/claims/${encodeURIComponent(id)}`)
  if (!res.ok) throw new Error('Failed to fetch claim')
  return (await res.json()) as Claim
}

export type Issue = {
  issue_type: string
  severity: string
  description: string
  evidence: Record<string, any>
}

export type ClaimAnalysis = {
  issues: Issue[]
  risk_score: number
  risk_level: string
}

export async function fetchClaimAnalysis(id: string) {
  const res = await fetch(`${BASE}/api/claims/${encodeURIComponent(id)}/analyze`)
  if (!res.ok) throw new Error('Failed to fetch analysis')
  return (await res.json()) as ClaimAnalysis
}

export type ApprovalStatus = 'pending' | 'approved' | 'declined' | 'escalated'

export type Recommendation = {
  id: number
  claim_id: number
  action_type: string | null
  rationale: string
  confidence: string | null
  low_confidence: boolean
  cited_issue_types: string[]
  secondary_options: { action_type: string; rationale: string; confidence: string }[]
  approval_status: ApprovalStatus
  created_at: string
  decided_at?: string | null
}

export type CommanderDecisionOut = {
  decision: string
  reason_code: string
  rule: number
}

export type ExecutionResult =
  | { kind: 'FollowUpResult'; claim_id: string; followup_id: number; note: string; due_at: string | null; attempts: number; activity_log_id: number }
  | { kind: 'FollowUpFailure'; claim_id: string | null; reason: string; detail: string; attempts: number; activity_log_id: number | null }
  | { kind: 'ReminderResult'; claim_id: string; reminder_id: number; target: string; content: string; reference_number: string | null; attempts: number; activity_log_id: number }
  | { kind: 'ReminderFailure'; claim_id: string | null; reason: string; detail: string; attempts: number; activity_log_id: number | null }

export type RecommendationActionResult = {
  recommendation: Recommendation
  decision: CommanderDecisionOut
  execution: ExecutionResult | null
  execution_decision: CommanderDecisionOut | null
  escalation_id: number | null
  activity?: ActivityLogEntry
}

export type GenerateRecommendationResult = {
  stage: 'pending' | 'escalated' | 'no_action'
  detail: string
  decision: CommanderDecisionOut | null
  recommendation: Recommendation | null
  escalation_id: number | null
}

export type ActivityLogEntry = {
  id: number
  action: string
  details: Record<string, any> | null
  created_at: string
}

async function asJson<T>(res: Response, errorMessage: string): Promise<T> {
  if (!res.ok) {
    let detail = errorMessage
    try {
      const body = await res.json()
      if (body?.detail) detail = body.detail
    } catch {
      // ignore — fall back to the generic message
    }
    throw new Error(detail)
  }
  return (await res.json()) as T
}

export async function fetchLatestRecommendation(claimId: string) {
  const res = await fetch(`${BASE}/api/claims/${encodeURIComponent(claimId)}/recommendation`)
  const body = await asJson<{ recommendation: Recommendation | null }>(res, 'Failed to fetch recommendation')
  return body.recommendation
}

export async function generateRecommendation(claimId: string) {
  const res = await fetch(`${BASE}/api/claims/${encodeURIComponent(claimId)}/recommendation`, { method: 'POST' })
  return asJson<GenerateRecommendationResult>(res, 'Failed to generate recommendation')
}

export async function approveRecommendation(claimId: string, recommendationId: number) {
  const res = await fetch(
    `${BASE}/api/claims/${encodeURIComponent(claimId)}/recommendation/${recommendationId}/approve`,
    { method: 'POST' },
  )
  return asJson<RecommendationActionResult>(res, 'Failed to approve recommendation')
}

export async function declineRecommendation(claimId: string, recommendationId: number) {
  const res = await fetch(
    `${BASE}/api/claims/${encodeURIComponent(claimId)}/recommendation/${recommendationId}/decline`,
    { method: 'POST' },
  )
  return asJson<RecommendationActionResult>(res, 'Failed to decline recommendation')
}

export async function fetchClaimActivity(claimId: string) {
  const res = await fetch(`${BASE}/api/claims/${encodeURIComponent(claimId)}/activity`)
  const body = await asJson<{ items: ActivityLogEntry[] }>(res, 'Failed to fetch activity')
  return body.items
}
